---
title: "BiocGF: interfacing to Geneformer with R"
author: "Vincent J. Carey, stvjc at channing.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{BiocGF: interfacing to Geneformer with R}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
bibliography: biocgf.bib
csl: acm-siggraph.csl
---

# Introduction

Geneformer is described in a Nature paper
entitled [Transfer learning enables predictions
in network biology](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10949956/)
@Theodoris2023.
The abstract concludes: "Overall, Geneformer represents a pretrained deep learning model from which fine-tuning towards a broad range of downstream applications can be pursued to accelerate discovery of key network regulators and candidate therapeutic targets."

Software resources related to the model are available at huggingface.co.

A preprint with critical content is [Kedzierska et al.](https://www.biorxiv.org/content/10.1101/2023.10.16.561085v2) @Kedzierska2023.  This package collects tools
and information related to both of these papers.

# Package resources

Programming for the analyses presented in the papers
is exclusively in python.  We use [basilisk](https://bioconductor.org/packages/basilisk) to collect specified versions of all Pypi-resident dependencies.
We also include a snapshot of the Geneformer software modules
with BiocGF.

## Exploring the model

The Huggingface model card for Geneformer indicates that the model
includes 10.3 million parameters.  Using the safetensors module we can
see the names of tensors of weights -- for example
```
> head(mkeys)
[1] "bert.embeddings.LayerNorm.bias"              
[2] "bert.embeddings.LayerNorm.weight"            
[3] "bert.embeddings.position_embeddings.weight"  
[4] "bert.embeddings.position_ids"                
[5] "bert.embeddings.token_type_embeddings.weight"
[6] "bert.embeddings.word_embeddings.weight"      
> tail(mkeys)
[1] "bert.encoder.layer.5.output.dense.weight"  
[2] "cls.predictions.bias"                      
[3] "cls.predictions.transform.LayerNorm.bias"  
[4] "cls.predictions.transform.LayerNorm.weight"
[5] "cls.predictions.transform.dense.bias"      
[6] "cls.predictions.transform.dense.weight"    
```

## A benchmark dataset

The Kedzierska paper includes a figshare link to a loom
file representing quantifications of single-cell RNA-seq
analyses of pancreas tissue.  This in turn refers to the
integration benchmarking paper of Luecken et al. @luecken,
including the remark:

```
We reprocessed the following public datasets for our integration tasks: pancreas
GSE81076, GSE85241, GSE86469, GSE84133, GSE81608 (Gene Expression
Omnibus (GEO)) and E-MTAB-5061 (ArrayExpress)...
```

This code chunk uses basilisk to hand back a python
reference to the loom representation of this dataset provided
by Kedzierska et al at [figshare](https://doi.org/10.6084/m9.figshare.24747228).  
This can be converted to SingleCellExperiment
in the chunk, before handing back to R.
```{r chkdat, message=FALSE}
library(BiocGF)
panc_path = get_panc_loom_path()
proc = basilisk::basiliskStart(BiocGF:::gfenv)
pancloom = basilisk::basiliskRun(proc, function() {
 anndata::read_loom(panc_path)
})
basilisk::basiliskStop(proc)
pancloom
```

This representation of the data is terse.  There is 
no metadata to discriminate donor or source (e.g., the GSE or MTAB
elements noted above).

## Tokenizing the loom RNA-seq content

The following call is time-consuming and is
not evaluated.

```{r tokcall,eval=FALSE}
tt = gf_tokenize_loom( panc_path )
```

A serialized version of the tokenization output
is available for this dataset.

```{r lktok}
cur = getwd()
td = tempdir()
setwd(td)
pa = get_panc_tok_path()
unzip(pa)  # destination = "."
ds = getgf() # get python module
ref = ds$load_from_disk("panc_hfds/tbyr.dataset")
print(ref)
setwd(cur)
```




# References
