<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BiocGF: interfacing to Geneformer with R • BiocGF</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BiocGF: interfacing to Geneformer with R">
<meta property="og:description" content="BiocGF">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocGF</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BiocGF.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.con/vjcitn/BiocGF/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="BiocGF_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BiocGF: interfacing to Geneformer with R</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">July 15, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.con/vjcitn/BiocGF/blob/HEAD/vignettes/BiocGF.Rmd" class="external-link"><code>vignettes/BiocGF.Rmd</code></a></small>
      <div class="hidden name"><code>BiocGF.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Geneformer is described in a Nature paper entitled <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10949956/" class="external-link">Transfer learning enables predictions in network biology</a> <span class="citation">Theodoris et al. [2023]</span>. The abstract concludes: “Overall, Geneformer represents a pretrained deep learning model from which fine-tuning towards a broad range of downstream applications can be pursued to accelerate discovery of key network regulators and candidate therapeutic targets.”</p>
<p>Software resources related to the model are available at huggingface.co.</p>
<p>A preprint with critical content is <a href="https://www.biorxiv.org/content/10.1101/2023.10.16.561085v2" class="external-link">Kedzierska et al.</a> <span class="citation">Kedzierska et al. [2023]</span>. This package collects tools and information related to both of these papers.</p>
</div>
<div class="section level2">
<h2 id="package-resources">Package resources<a class="anchor" aria-label="anchor" href="#package-resources"></a>
</h2>
<p>Programming for the analyses presented in the papers is exclusively in python. We use <a href="https://bioconductor.org/packages/basilisk" class="external-link">basilisk</a> to collect specified versions of all Pypi-resident dependencies. We also include a snapshot of the Geneformer software modules with BiocGF.</p>
<div class="section level3">
<h3 id="exploring-the-model">Exploring the model<a class="anchor" aria-label="anchor" href="#exploring-the-model"></a>
</h3>
<p>The Huggingface model card for Geneformer indicates that the model includes 10.3 million parameters. Using the safetensors module we can see the tensors of weights. <code>get_gf_safetensors</code> retrieves an image of model weights distributed in the huggingface repository for Geneformer.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.con/vjcitn/BiocGF" class="external-link">BiocGF</a></span><span class="op">)</span></span>
<span><span class="va">pa</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_gf_safetensors.html">get_gf_safetensors</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">td</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">pa</span>, exdir<span class="op">=</span><span class="va">td</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/nwts_safet.html">nwts_safet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="st">"GF_safetensors"</span>, <span class="st">"model.safetensors"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10290770</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">st</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_safet.html">get_safet</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># not standard, hands live python ref</span></span>
<span><span class="va">oo</span> <span class="op">=</span> <span class="va">st</span><span class="op">$</span><span class="fu">safe_open</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="st">"GF_safetensors"</span>, <span class="st">"model.safetensors"</span><span class="op">)</span>, <span class="st">"pt"</span><span class="op">)</span></span>
<span><span class="va">mk</span> <span class="op">=</span> <span class="va">oo</span><span class="op">$</span><span class="fu">keys</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mk</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 107</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">mk</span>, <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="va">oo</span><span class="op">$</span><span class="fu">get_tensor</span><span class="op">(</span><span class="va">z</span><span class="op">)</span><span class="op">$</span><span class="va">shape</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $bert.embeddings.LayerNorm.bias</span></span>
<span><span class="co">## torch.Size([256])</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bert.embeddings.LayerNorm.weight</span></span>
<span><span class="co">## torch.Size([256])</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bert.embeddings.position_embeddings.weight</span></span>
<span><span class="co">## torch.Size([2048, 256])</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bert.embeddings.position_ids</span></span>
<span><span class="co">## torch.Size([1, 2048])</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bert.embeddings.token_type_embeddings.weight</span></span>
<span><span class="co">## torch.Size([2, 256])</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bert.embeddings.word_embeddings.weight</span></span>
<span><span class="co">## torch.Size([25426, 256])</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="a-benchmark-dataset">A benchmark dataset<a class="anchor" aria-label="anchor" href="#a-benchmark-dataset"></a>
</h3>
<p>The Kedzierska paper includes a figshare link to a loom file representing quantifications of single-cell RNA-seq analyses of pancreas tissue. This in turn refers to the integration benchmarking paper of Luecken et al. <span class="citation">Luecken et al. [2022]</span>, including the remark:</p>
<pre><code>We reprocessed the following public datasets for our integration tasks: pancreas
GSE81076, GSE85241, GSE86469, GSE84133, GSE81608 (Gene Expression
Omnibus (GEO)) and E-MTAB-5061 (ArrayExpress)...</code></pre>
<p>This code chunk uses basilisk to hand back a python reference to the loom representation of this dataset provided by Kedzierska et al at <a href="https://doi.org/10.6084/m9.figshare.24747228" class="external-link">figshare</a>.<br>
This can be converted to SingleCellExperiment in the chunk, before handing back to R.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.con/vjcitn/BiocGF" class="external-link">BiocGF</a></span><span class="op">)</span></span>
<span><span class="va">panc_path</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_panc_loom_path.html">get_panc_loom_path</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">proc</span> <span class="op">=</span> <span class="fu">basilisk</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/basilisk/man/basiliskStart.html" class="external-link">basiliskStart</a></span><span class="op">(</span><span class="fu">BiocGF</span><span class="fu">:::</span><span class="va">gfenv</span><span class="op">)</span></span>
<span><span class="va">pancloom</span> <span class="op">=</span> <span class="fu">basilisk</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/basilisk/man/basiliskStart.html" class="external-link">basiliskRun</a></span><span class="op">(</span><span class="va">proc</span>, <span class="kw">function</span><span class="op">(</span><span class="op">)</span> <span class="op">{</span></span>
<span> <span class="fu">anndata</span><span class="fu">::</span><span class="fu"><a href="https://anndata.dynverse.org/reference/read_loom.html" class="external-link">read_loom</a></span><span class="op">(</span><span class="va">panc_path</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">basilisk</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/basilisk/man/basiliskStart.html" class="external-link">basiliskStop</a></span><span class="op">(</span><span class="va">proc</span><span class="op">)</span></span>
<span><span class="va">pancloom</span></span></code></pre></div>
<pre><code><span><span class="co">## AnnData object with n_obs × n_vars = 16382 × 16388</span></span>
<span><span class="co">##     obs: 'adata_order', 'batch', 'celltype', 'n_counts', 'n_genes', 'n_genes_by_counts', 'obs_names', 'size_factors', 'tech', 'total_counts'</span></span>
<span><span class="co">##     var: 'ensembl_id', 'gene_symbols', 'has_ensembl_match', 'mean_counts', 'n_cells', 'n_cells_by_counts', 'pct_dropout_by_counts', 'total_counts', 'var_names'</span></span>
<span><span class="co">##     layers: 'counts'</span></span></code></pre>
<p>This representation of the data is terse. There is no metadata to discriminate donor or source (e.g., the GSE or MTAB elements noted above).</p>
</div>
<div class="section level3">
<h3 id="tokenizing-the-loom-rna-seq-content">Tokenizing the loom RNA-seq content<a class="anchor" aria-label="anchor" href="#tokenizing-the-loom-rna-seq-content"></a>
</h3>
<p>The following call is time-consuming and is not evaluated.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tt</span> <span class="op">=</span> <span class="fu"><a href="../reference/gf_tokenize_loom.html">gf_tokenize_loom</a></span><span class="op">(</span> <span class="va">panc_path</span> <span class="op">)</span></span></code></pre></div>
<p>A serialized version of the tokenization output is available for this dataset.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cur</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">td</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">td</span><span class="op">)</span></span>
<span><span class="va">pa</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_panc_tok_path.html">get_panc_tok_path</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">pa</span><span class="op">)</span>  <span class="co"># destination = "."</span></span>
<span><span class="va">ds</span> <span class="op">=</span> <span class="fu"><a href="../reference/getgf.html">getgf</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># get python module</span></span>
<span><span class="va">ref</span> <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="fu">load_from_disk</span><span class="op">(</span><span class="st">"panc_hfds/tbyr.dataset"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ref</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Dataset({</span></span>
<span><span class="co">##     features: ['input_ids', 'length'],</span></span>
<span><span class="co">##     num_rows: 16382</span></span>
<span><span class="co">## })</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">cur</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BiocGF_0.0.3     BiocStyle_2.32.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.9            utf8_1.2.4            generics_0.1.3       </span></span>
<span><span class="co">##  [4] lattice_0.22-6        RSQLite_2.3.7         digest_0.6.36        </span></span>
<span><span class="co">##  [7] magrittr_2.0.3        grid_4.4.1            evaluate_0.24.0      </span></span>
<span><span class="co">## [10] bookdown_0.40         fastmap_1.2.0         blob_1.2.4           </span></span>
<span><span class="co">## [13] jsonlite_1.8.8        Matrix_1.7-0          anndata_0.7.5.6      </span></span>
<span><span class="co">## [16] DBI_1.2.3             BiocManager_1.30.23   httr_1.4.7           </span></span>
<span><span class="co">## [19] purrr_1.0.2           fansi_1.0.6           textshaping_0.4.0    </span></span>
<span><span class="co">## [22] jquerylib_0.1.4       cli_3.6.3             rlang_1.1.4          </span></span>
<span><span class="co">## [25] dbplyr_2.5.0          basilisk.utils_1.16.0 bit64_4.0.5          </span></span>
<span><span class="co">## [28] withr_3.0.0           cachem_1.1.0          yaml_2.3.8           </span></span>
<span><span class="co">## [31] parallel_4.4.1        dir.expiry_1.12.0     tools_4.4.1          </span></span>
<span><span class="co">## [34] memoise_2.0.1         dplyr_1.1.4           filelock_1.0.3       </span></span>
<span><span class="co">## [37] basilisk_1.16.0       reticulate_1.38.0     assertthat_0.2.1     </span></span>
<span><span class="co">## [40] curl_5.2.1            png_0.1-8             vctrs_0.6.5          </span></span>
<span><span class="co">## [43] R6_2.5.1              BiocFileCache_2.12.0  lifecycle_1.0.4      </span></span>
<span><span class="co">## [46] fs_1.6.4              htmlwidgets_1.6.4     bit_4.0.5            </span></span>
<span><span class="co">## [49] ragg_1.3.2            pkgconfig_2.0.3       desc_1.4.3           </span></span>
<span><span class="co">## [52] pkgdown_2.0.9         bslib_0.7.0           pillar_1.9.0         </span></span>
<span><span class="co">## [55] Rcpp_1.0.12           glue_1.7.0            systemfonts_1.1.0    </span></span>
<span><span class="co">## [58] xfun_0.45             tibble_3.2.1          tidyselect_1.2.1     </span></span>
<span><span class="co">## [61] knitr_1.47            htmltools_0.5.8.1     rmarkdown_2.27       </span></span>
<span><span class="co">## [64] compiler_4.4.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references hanging-indent">
<div id="ref-Kedzierska2023">
<p><span class="smallcaps">Kedzierska, K.Z., Crawford, L., Amini, A.P., and Lu, A.X.</span> 2023. Assessing the limits of zero-shot foundation models in single-cell biology. <em>bioRxiv</em>.</p>
</div>
<div id="ref-Luecken2022">
<p><span class="smallcaps">Luecken, M.D., Büttner, M., Chaichoompu, K., et al.</span> 2022. Benchmarking atlas-level data integration in single-cell genomics. <em>Nature Methods</em> <em>19</em>, 1, 41–50.</p>
</div>
<div id="ref-Theodoris2023">
<p><span class="smallcaps">Theodoris, C.V., Xiao, L., Chopra, A., et al.</span> 2023. Transfer learning enables predictions in network biology. <em>Nature</em> <em>618</em>, 7965, 616–624.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vince Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
